cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -149,9 +150,22 @@ class ParquetScanTaskIterator {
   static Status InferColumnProjection(const parquet::FileMetaData& metadata,
                                       const std::shared_ptr<ScanOptions>& options,
                                       std::vector<int>* out) {
-    // TODO(fsaintjacques): Compute intersection _and_ validity
-    *out = internal::Iota(metadata.num_columns());
+    if (options->projector == nullptr || options->schema == nullptr) {
+      *out = internal::Iota(metadata.num_columns());
+      return Status::OK();
+    }
 
+    // union fields in expression and fields in projector
+    auto projector_schema = options->projector->schema();
+    auto filter_schema =
+        SchemaFromColumnNames(options->schema, FieldsInExpression(options->filter));
+    ARROW_ASSIGN_OR_RAISE(auto merged, MergeSchemas({projector_schema, filter_schema}));
+
+    // get column indices
+    out->clear();
+    for (const auto& field : merged->fields()) {
+      out->push_back(options->schema->GetFieldIndex(field->name()));
```

Comment:
Alright, I'll push projection inference down to row group level and use each row group's schema.

Fields absent from the row group's schema will be added by the projector as a column of nulls. If they are referenced by the filter then that field will be null at predicate pushdown time and the simplified filter expression will not reference it anymore. I'll add tests for these

Reply:
Relying on the optimizer to elide the field is a brittle and false assumption to make, e.g. the following filter will almost always pass even if the column is missing `IF(RAND(1, 10000) < 2, missing_col > 2, true)`. We need to account (and not fail?) for this in the column selection.

It make me realize that we sometimes have to inject a projector _before_ the filter. This will be part of the Schema normalization sub-project.

Reply:
I'm not sure how this is brittle; that case would simplify as follows:
```
If(Rand(1, 10000) < 2, "missing_col"_ > 2, true)
If(Rand(1, 10000) < 2, null > 2, true)
If(Rand(1, 10000) < 2, null, true)
```